terraform/charm/large_deployment/main.tf

```diff
@@ -0,0 +1,185 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+locals {
+  apps = [
+    for app in concat(var.apps != null ? var.apps : []) : app if app != null
+  ]
+
+  apps_in_main_model = [
+    for app in concat([var.failover], local.apps) :
+    app if app != null && app.model == var.main.model
+  ]
+  apps_not_in_main_model = [
+    for app in concat([var.failover], local.apps) :
+    app if app != null && app.model != var.main.model
+  ]
+  apps_in_failover_model = [
+    for app in local.apps :
+    app if app.model == var.failover.model
+  ]
+  apps_not_in_failover_model = [
+    for app in local.apps :
+    app if app.model != var.failover.model
+  ]
+
+  all_models = distinct(concat(
```

Comment:
I don't see this as a blocker but we can offer relations and consume on the same model, e.g.:
```
juju add-model test
juju deploy kafka --channel=3/edge --config profile=testing
juju deploy zookeeper --channel=3/edge
```

Then we can offer the ZK relation and consume it locally:
```
juju offer zookeeper:zookeeper zookeeper-client
juju relate kafka admin/test.zookeeper-client
```

Which results in: https://pastebin.ubuntu.com/p/XF62dDq5VW/

This way, all key relations get offered and we consume them on the models we need.

Reply:
Re-reading the code, I believe that is what you've done with `all_models > 1`?

Reply:
Yeah, I would like to avoid offering opensearch or self-signed-certificates unless explicitly deployed in cross models from the get go